### PR TITLE
OpenTelemetry: propagate a configured context(s) to root requests

### PR DESCRIPTION
**Pull request by dvoytenko**
Implement OTEL propagation according to the OpenTelemetry API per https://opentelemetry.io/docs/specs/otel/context/api-propagators/. Any configured propagator should work with this API, including `W3CTraceContextPropagator` (https://www.w3.org/TR/trace-context/).

Alternative to the https://github.com/vercel/next.js/pull/56891.

/cc @sfirrin

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| buildDuration | 10.6s | 10.5s | N/A |
| buildDurationCached | 6.1s | 6.1s | N/A |
| nodeModulesSize | 174 MB | 174 MB | ⚠️ +3.82 kB |
| nextStartRea..uration (ms) | 520ms | 510ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| 199-HASH.js gzip | 27.5 kB | 27.5 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.3 kB | 45.3 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 32.9 kB | 32.9 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.8 kB | 98.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | N/A |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.36 kB | 4.36 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.63 kB | N/A |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.08 kB | 1.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| index.html gzip | 528 B | 530 B | N/A |
| link.html gzip | 541 B | 542 B | N/A |
| withRouter.html gzip | 523 B | 524 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.7 kB | 93.8 kB | N/A |
| page.js gzip | 155 kB | 155 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | dvoytenko/next.js opentelemetry/propagate | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 622 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 22.5 kB | 22.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 22bc5f8adb712f7cb8f7ae3d38732e3112c3e6d8

File: packages/next/src/server/lib/trace/tracer.ts [hunk 1]
```diff
@@ -28,7 +29,8 @@ if (process.env.NEXT_RUNTIME === 'edge') {
   }
 }
 
-const { context, trace, SpanStatusCode, SpanKind } = api
+const { context, propagation, trace, SpanStatusCode, SpanKind, ROOT_CONTEXT } =
```
Comment (agadzik) [inline packages/next/src/server/lib/trace/tracer.ts:32]: Using ROOT_CONTEXT will break when we roll auto registration. Can you use context.active() instead?

**Review by agadzik — CHANGES_REQUESTED** [review]

File: packages/next/src/server/lib/trace/tracer.ts (hunk 1 above)
Comment (dvoytenko) [reply packages/next/src/server/lib/trace/tracer.ts:32]: To change `ROOT_CONTEXT`, this code would have to change: https://github.com/vercel/next.js/blob/254d6d5d6d4cd74ae539313939ee9e994fa9a8d5/packages/next/src/server/lib/trace/tracer.ts#L231-L234

It's probably the right thing to do overall, but would be a traumatic change here. IMHO we should take that on separately from this PR.

Comment (dvoytenko) [reply packages/next/src/server/lib/trace/tracer.ts:32]: @agadzik I thought about this a bit more and I actually think using `context.active()` wouldn't be right here. IMHO, if we allow any infra to create parent spans before getting to Next's handler, that infra should be invoking propagation.

Comment (dvoytenko) [reply packages/next/src/server/lib/trace/tracer.ts:32]: @agadzik I come up with a simple compromise. The propagation is only done on the root context now. This way it should be always correct in one general case and we can always add support for non-root contexts as we see need later.

**Comment by mplachter** [comment]
Is there any update on this being merged and going out?

**Review by feedthejim — APPROVED** [review]

**Pull request merged**